packages/block-editor/src/components/block-preview/index.js

```diff
@@ -33,19 +37,57 @@ function BlockPreview( props ) {
 }
 
 export function BlockPreviewContent( { blocks, settings } ) {
+	const previewRef = useRef( null );
+
+	const [ previewScale, setPreviewScale ] = useState( 1 );
+	const [ visibility, setVisibility ] = useState( 'hidden' );
+
+	let timerId;
+
+	// Dynamically calculate the scale factor
+	useLayoutEffect( () => {
+		const { clientId } = blocks;
+		timerId = setTimeout( () => {
+			const refNode = previewRef.current;
+			const containerWidth = refNode.offsetWidth - ( 14 * 2 );
+
+			const previewDomElements = getClockPreviewContainerDOMNode( clientId );
+			const previewBlocksWidth = previewDomElements ? previewDomElements.offsetWidth : containerWidth;
+			const scale = Math.min( containerWidth / previewBlocksWidth ) || 1;
+
+			setPreviewScale( scale );
+			setVisibility( 'visible' );
+		}, 10 );
+	} );
+
+	useEffect( () => {
+		return () => window.clearTimeout( timerId );
```

Comment:
Why are you clearing the timer here and not in the layout effect? Genuinely curious.

I did this by returning a function from the `useLayoutEffect` which cleared the timer.

Reply:
I've fixed this one.